File: Phoenix-SDK-June-2008-RC1/SourceDir/Phoenix SDK June 2008/samples/Pascal/cpp/FrontEnd/Output.cpp
//-----------------------------------------------------------------------------
//
// Phoenix
// Copyright (C) Microsoft Corporation.  All Rights Reserved.
//
// Description:
//
//    Implementation of the Output class.
//
//-----------------------------------------------------------------------------

using namespace System;
using namespace System::Diagnostics;

#include "stdafx.h"
#include "Ast.h"

namespace Pascal
{

TextWriter ^ 
Output::DefaultWriter::get()
{
   return System::Console::Out;
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportMessage
(
   String ^ message
)
{
   ReportMessage(-1, message);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportMessage
(
   int lineNumber, 
   String ^ message
)
{
   // Format message string.

   if (lineNumber >= 0)
   {
      message = String::Format(
         "{0}({1}) : {2}",
         CurrentSourceFileName, 
         lineNumber, 
         message
      );
   }
   else
   {
      message = String::Format(
         "{0}",
         message);
   }

   // Write the message to the message writer if one was specified;
   // otherwise, write the message to the console.

   if (MessageWriter)
   {
      MessageWriter->WriteLine(message);
   }
   else
   {
      Console::ForegroundColor = System::ConsoleColor::Cyan;
      Console::WriteLine(message);
      Console::ResetColor();
   }
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given warning message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportWarning
(     
   int lineNumber,
   Error warning, 
   Object ^ arg0
)
{
   String ^ message = String::Format(errorMap[warning], arg0);
   InternalReportMessage(Output::warning, lineNumber, message, (int)warning);
   WarningCount++;
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportError
(
   int lineNumber, 
   Error error
)
{
   InternalReportMessage(Output::error, lineNumber, 
      errorMap[error], (int)error);
   ErrorCount++;
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportError
(
   int lineNumber, 
   Error error,
   Object ^ arg0
)
{
   String ^ message = String::Format(errorMap[error], arg0);
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportError
(
   int lineNumber, 
   Error error,
   Object ^ arg0, 
   Object ^ arg1
)
{
   String ^ message = String::Format(errorMap[error], arg0, arg1);
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportError
(
   int lineNumber, 
   Error error,
   Object ^ arg0, 
   Object ^ arg1, 
   Object ^ arg2
)
{
   String ^ message = String::Format(errorMap[error], arg0, arg1, arg2);
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given fatal error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportFatalError(Error error)
{
   ReportFatalError(-1, error);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given fatal error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportFatalError(Error error, Object ^ arg0)
{
   ReportFatalError(-1, error, arg0);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given fatal error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void Output::ReportFatalError
(
   int lineNumber, 
   Error error
)
{
   String ^ message = errorMap[error];
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;

   InternalReportMessage(Output::fatalError, lineNumber, 
      "unable to recover from previous error(s); stopping compilation.",
      (int)Error::FatalError);

   // Throw fatal error message. This will be caught by the main() function.

   throw gcnew FatalErrorException(lineNumber, message);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given fatal error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void Output::ReportFatalError
(
   int lineNumber, 
   Error error, 
   Object ^ arg0
)
{
   String ^ message = String::Format(errorMap[error], arg0);
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;

   InternalReportMessage(Output::fatalError, lineNumber, 
      "unable to recover from previous error(s); stopping compilation.", 
      (int)Error::FatalError);

   // Throw fatal error message. This will be caught by the main() function.

   throw gcnew FatalErrorException(lineNumber, message);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given fatal error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportFatalError
(
   int lineNumber,
   Error error,
   Object ^ arg0,
   Object ^ arg1
)
{
   String ^ message = String::Format(errorMap[error], arg0, arg1);
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;

   InternalReportMessage(Output::fatalError, lineNumber, 
      "unable to recover from previous error(s); stopping compilation.",
      (int)Error::FatalError);

   // Throw fatal error message. This will be caught by the main() function.

   throw gcnew FatalErrorException(lineNumber, message);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Writes the given fatal error message to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void 
Output::ReportFatalError
(
   int lineNumber,
   Error error,
   Object ^ arg0,
   Object ^ arg1,
   Object ^ arg2
)
{
   String ^ message = String::Format(errorMap[error], arg0, arg1, arg2);
   InternalReportMessage(Output::error, lineNumber, message, (int)error);
   ErrorCount++;

   InternalReportMessage(Output::fatalError, lineNumber, 
      "unable to recover from previous error(s); stopping compilation.",
      (int)Error::FatalError);

   // Throw fatal error message. This will be caught by the main() function.

   throw gcnew FatalErrorException(lineNumber, message);
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Internal method for writing strings to the output stream.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void
Output::InternalReportMessage
(
   String ^ messageType, 
   int      lineNumber,
   String ^ message,
   int      messageNumber
)
{
   String ^ output;

   // Write source file / line information if a valid line number
   // was provided.

   if (lineNumber >= 0)
   {
      output = String::Format(
         "{0}({1}) ",
         CurrentSourceFileName, 
         lineNumber
      );

      if (MessageWriter)
         MessageWriter->Write(output);
      else
         Console::Write(output);
   }

   // Add in the message number if one was provided.

   if (messageNumber >= 0)
      output = String::Format("{0} P{1}", messageType, messageNumber);
   else
      output = messageType;

   // Write the output.

   if (MessageWriter)
      MessageWriter->Write(output);
   else
   {      
      if (messageType->Equals(error) || messageType->Equals(fatalError))
      {
         Console::ForegroundColor = System::ConsoleColor::Red;
      }
      Console::Write(output);
      Console::ResetColor();
   }

   output = String::Format(": {0}", message);

   if (MessageWriter)
      MessageWriter->WriteLine(output);
   else
   {      
      Console::WriteLine(output);
   }
}

//-----------------------------------------------------------------------------
//
// Description:
//
//    Initializes the error map.
//
// Remarks:
//
//
// Returns:
//
//   void
//
//-----------------------------------------------------------------------------

void
Output::InitializeErrorMap()
{
   errorMap = gcnew Dictionary<Error, String ^>();

   errorMap[Error::EmptyFileList] = 
   "no file name(s) provided.";

   errorMap[Error::InvalidFile] = 
   "cannot open file {0}.";

   errorMap[Error::ProcessNotFound] = 
   "Failed to launch '{0}'{2}Reason: '{1}'{2}"
   "Ensure the application exists and is accessible from your system "
   "PATH environment variable.";

   errorMap[Error::InternalCompilerError] = 
   "{0}{2}Stack Trace:{2}{1}";

   errorMap[Error::UnknownInternalCompilerError] = 
   "Internal Compiler Error";

   errorMap[Error::InvalidEscapeSequence] = 
   "invalid escape sequence.";

   errorMap[Error::ParseSyntaxError] = 
   "syntax error: token '{0}'.";

   errorMap[Error::UnknownIndexType] = 
   "'{0}': unknown index type; expected scalar or subrange type.";

   errorMap[Error::FileArrayUnsupported] =
   "array of files is unsupported.";

   errorMap[Error::ZeroIndexRange] = 
   "index type range must be non-zero.";

   errorMap[Error::NegativeIndexRange] = 
   "index type range must be non-negative.";

   errorMap[Error::BadVariableAssignment] = 
   "variable '{0}' of type '{1}' is being assigned an "
   "expression of type '{2}'.";

   errorMap[Error::ConstantAssignment] = 
   "'{0}': you cannot assign to a variable that is constant.";

   errorMap[Error::FileAssignment] = 
   "'{0}': you cannot assign to a file variable.";

   errorMap[Error::FunctionAssignment] = 
   "'{0}': you cannot assign to a function or procedure.";

   errorMap[Error::GeneralTypeExpected] = 
   "expected {0} type."; 

   errorMap[Error::IncompatibleCaseIndexType] =
   "type '{0}' is not compatible with case index type '{1}'.";

   errorMap[Error::CaseStatementRange] =
   "'{0}..{1}': range already appears in case statement list.";
   
   errorMap[Error::CaseStatementValue] =
   "'{0}': value already appears in case statement list.";

   errorMap[Error::UnknownType] = 
   "type '{0}' not found.";

   errorMap[Error::TypeRedefinition] =
   "'{0}': type redefinition.";

   errorMap[Error::RangeRedefinition] =
   "redefinition of range for type '{0}'; ({1}) -> ({2}).";

   errorMap[Error::StaticRangeCheckFailure] =
   "value '{0}' exceeded legal bounds ({1}..{2}).";

   errorMap[Error::UnknownRange] = 
   "internal error: compiler could not find range mapping for '{0}'.";

   errorMap[Error::IncompatibleTypes] =
   "incompatible types.";

   errorMap[Error::IncompatibleOperandTypes] =
   "'{0} {1} {2}': operand types are not compatible.";

   errorMap[Error::InvalidPointerAccess] =
   "expected pointer access symbol '^' or '->'.";

   errorMap[Error::InvalidFieldAccess] =
   "'{0}' is not a member of '{1}'.";

   errorMap[Error::UndeclaredLabel] =
   "'{0}': undeclared label.";

   errorMap[Error::UndefinedLabel] =
   "'{0}': undefined label.";

   errorMap[Error::LabelRedeclaration] =
   "label '{0}' already declared.";

   errorMap[Error::LabelRedefinition] =
   "'{0}': label refinition.";

   errorMap[Error::InvalidLabelLength] =
   "'{0}': labels may contain at most 4 digits.";

   errorMap[Error::DeclaredUndefinedLabel] = 
   "label '{0}' was declared but never defined.";

   errorMap[Error::UnreferencedLabel] = 
   "label '{0}' was declared but never referenced.";

   errorMap[Error::ZeroSizeRecord] = 
   "zero-sized records are not allowed.";

   errorMap[Error::RecursiveRecordDefinition] = 
   "'{0}.{0}' : Recursive definitions of records are not allowed.";

   errorMap[Error::FieldRedeclaration] = 
   "'{0}' already declared as field with type '{1}'.";

   errorMap[Error::UnknownSetBaseType] =
   "unknown set base type.";

   errorMap[Error::InvalidSetType] =
   "'{0}': invalid set type; expected scalar or subrange type.";

   errorMap[Error::MismatchedSubrangeLimit] =
   "subrange limit types do not match ('{0}', '{1}').";

   errorMap[Error::InvalidSubrange] =
   "invalid subrange ({0}..{1}).";

   errorMap[Error::UndeclaredIdentifier] =
   "'{0}': undeclared identifier.";

   errorMap[Error::AmbiguousFileAccess] = 
   "'{0}': ambiguous file access.";

   errorMap[Error::IncompatibleSetTypes] = 
   "incompatible set types.";

   errorMap[Error::IncompatibleSetTypesForMembership] = 
   "incompatible types for set membership.";

   errorMap[Error::MismatchedMemberDesignators] = 
   "mismatched types in member designator list.";

   errorMap[Error::IndexListCount] =
   "index list count mismatch ({0}, {1}).";

   errorMap[Error::MismatchedIndexType] = 
   "mismatched index type (expected '{0}').";

   errorMap[Error::MismatchedArrayElementTypes] =
   "'{0}': mismatched array element types.";

   errorMap[Error::MismatchedArraySizes] =
   "'{0}': mismatched array sizes.";

   errorMap[Error::InvalidWithField] =
   "'{0}' : invalid field for with clause.";

   errorMap[Error::FormalParameterRedeclaration] =
   "redeclaration of formal parameter '{0}'.";

   errorMap[Error::ProgramParameterRedeclaration] =
   "redeclaration of program parameter '{0}'.";

   errorMap[Error::UndeclaredProgramParameter] = 
   "file '{0}' must be declared in program parameter list.";

   errorMap[Error::MustBeVarParameter] = 
   "'{0}': {1} parameters must be specified as var-parameters.";

   errorMap[Error::ConstantVarParameter] = 
   "'{0}': var-parameter '{1}' must be non-constant.";

   errorMap[Error::InvalidArgument] = 
   "'{0}': expected '{1}' type for argument {2}.";

   errorMap[Error::InvalidArgumentCount] = 
   "'{0}': expected {1} arguments; {2} provided.";

   errorMap[Error::InvalidProcedureArgument] = 
   "procedure and function arguments may only contain value parameters.";

   errorMap[Error::InvalidProcedureArgument] = 
   "procedure and function arguments may only contain value parameters.";

   errorMap[Error::SymbolRedeclaration] = 
   "'{0}' already declared as type '{1}'.";

   errorMap[Error::FunctionRedeclaration] =
   "there is already a {0} named '{1}'.";

   errorMap[Error::InvalidReturnType] =
   "function return type must be scalar, subrange, or pointer type.";

   errorMap[Error::FunctionUnreachable] = 
   "'{0}': {1} not reachable from current scope.";

   errorMap[Error::UndeclaredFunction] = 
   "'{0}': undeclared procedure or function.";

   errorMap[Error::UnknownDirective] = 
   "'{0}': unknown directive.";

   errorMap[Error::InvalidExternalFunctionDefinition] = 
   "'{0}': definition of external {1} must occur in separate module.";

   errorMap[Error::InvalidForwardFunctionDefinition] = 
   "'{0}': {1} is forward-declared; repetition of parameter list not allowed.";

   errorMap[Error::UnknownTypeSymbol] = 
   "Could not resolve symbol for type '{0}'";

   errorMap[Error::InvalidWidthSpecifier] = 
   "invalid width specifier.";
}

}  // namespace Pascal

File: Phoenix-SDK-June-2008-RC1/SourceDir/Phoenix SDK June 2008/samples/Lispkit/csharp/Output.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Lispkit
{
   /// <summary>
   /// The exception that is thrown when a fatal compiler error occurs. 
   /// The compiler throws this error when it encounters an error that 
   /// it cannot recover from.
   /// </summary>
   class FatalErrorException : System.Exception
   {
      private int lineNumber;

      // Retrieves the source line number that is associated with the 
      // fatal error.
      public int LineNumber
      {
         get { return this.lineNumber; }
      }

      public FatalErrorException(int lineNumber, string message)
         :  base(message)
      {
         this.lineNumber = lineNumber;
      }     
   };

   /// <summary>
   /// Defines all errors and warnings for the compiler.
   /// </summary>
   enum Error
   {
      FatalError = 1000,            // Used to report fatal errors.
      InternalError = 1001,         // Used to report an internal error 
                                    // in the compiler.      
      FileNotFound = 2000,          // File not found error.
      TooManyArguments = 2001,      // Used to report cases where too many 
                                    // arguments are passed to a higher-order
                                    // function.
      UndeclaredIdentifier = 2002,  // Used to report undeclared identifiers.
   }
      
   /// <summary>
   /// Writes error and warning diagnostic messages to an output stream.
   /// </summary>
   static class Output
   {
      // Writes the given message to the output stream.
      internal static void ReportMessage(string message)
      {
         ReportMessage(-1, message);
      }

      // Writes the given message to the output stream.
      internal static void ReportMessage(int lineNumber, string message)
      {
         // Format message string.
         if (lineNumber >= 0)
         {
            message = string.Format("{0}({1}) : {2}", 
               SourceFileName, lineNumber, message);
         }
         else
         {
            message = string.Format("{0}", message);
         }

         // Write the message to the message writer if one was specified;
         // otherwise, write the message to the console.

         if (MessageWriter != null)
         {
            MessageWriter.WriteLine(message);
         }
         else
         {
            Console.ForegroundColor = System.ConsoleColor.Cyan;
            Console.WriteLine(message);
            Console.ResetColor();
         }
      }

      // Writes the given message to the output stream if trace
      // mode if enabled.
      internal static void TraceMessage()
      {
         TraceMessage(string.Empty);
      }

      // Writes the given message to the output stream if trace
      // mode if enabled.
      internal static void TraceMessage(string message)
      {
         TraceMessage(-1, message);
      }

      // Writes the given message to the output stream if trace
      // mode if enabled.
      internal static void TraceMessage(int lineNumber, string message)
      {
         if (!Program.VerboseMode)
         {
            return;
         }

         if (message.Length == 0)
         {
            if (MessageWriter != null)
            {
               MessageWriter.WriteLine();
            }
            else
            {
               Console.WriteLine();
            }
            return;
         }

         // Format message string.
         if (lineNumber >= 0)
         {
            message = string.Format("{0}({1}) : {2}",
               SourceFileName, lineNumber, message);
         }
         else
         {
            message = string.Format("{0}", message);
         }

         // Write the message to the message writer if one was specified;
         // otherwise, write the message to the console.

         if (MessageWriter != null)
         {
            MessageWriter.WriteLine(message);
         }
         else
         {
            Console.ForegroundColor = System.ConsoleColor.Yellow;
            Console.WriteLine(string.Format("{0}: {1}", trace, message));
            Console.ResetColor();
         }
      }
      
      // Writes the given warning message to the output stream.
      internal static void ReportWarning(int lineNumber, Error warning, 
         object arg0)
      {
         string message = string.Format(errorMap[warning], arg0);
         InternalReportMessage(Output.warning, lineNumber, message, 
            (int)warning);
         WarningCount++;
      }

      // Writes the given error message to the output stream.
      internal static void ReportError(int lineNumber, Error error)
      {
         InternalReportMessage(Output.error, lineNumber, 
            errorMap[error], (int)error);
         ErrorCount++;
      }

      // Writes the given error message to the output stream.
      internal static void ReportError(int lineNumber, Error error, 
         object arg0)
      {
         string message = string.Format(errorMap[error], arg0);
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;
      }

      // Writes the given error message to the output stream.
      internal static void ReportError(int lineNumber, Error error, 
         object arg0, object arg1)
      {
         string message = string.Format(errorMap[error], arg0, arg1);
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;
      }

      // Writes the given error message to the output stream.
      internal static void ReportError(int lineNumber, Error error, 
         object arg0, object arg1, object arg2)
      {
         string message = string.Format(errorMap[error], arg0, arg1, arg2);
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;
      }

      // Writes the given fatal error message to the output stream.
      internal static void ReportFatalError(Error error)
      {
         ReportFatalError(-1, error);
      }

      // Writes the given fatal error message to the output stream.
      internal static void ReportFatalError(Error error, object arg0)
      {
         ReportFatalError(-1, error, arg0);
      }

      // Writes the given fatal error message to the output stream.
      internal static void ReportFatalError(int lineNumber, Error error)
      {
         string message = errorMap[error];
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;

         InternalReportMessage(Output.fatalError, lineNumber, 
            "unable to recover from previous error(s); stopping compilation.",
            (int)Error.FatalError);

         // Throw fatal error message.
         throw new FatalErrorException(lineNumber, message);
      }

      // Writes the given fatal error message to the output stream.
      internal static void ReportFatalError(int lineNumber, Error error, 
         object arg0)
      {
         string message = string.Format(errorMap[error], arg0);
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;

         InternalReportMessage(Output.fatalError, lineNumber, 
            "unable to recover from previous error(s); stopping compilation.", 
            (int)Error.FatalError);

         // Throw fatal error message.
         throw new FatalErrorException(lineNumber, message);
      }

      // Writes the given fatal error message to the output stream.
      internal static void ReportFatalError(int lineNumber, Error error, 
         object arg0, object arg1)
      {
         string message = string.Format(errorMap[error], arg0, arg1);
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;

         InternalReportMessage(Output.fatalError, lineNumber, 
            "unable to recover from previous error(s); stopping compilation.",
            (int)Error.FatalError);

         // Throw fatal error message.
         throw new FatalErrorException(lineNumber, message);
      }

      // Writes the given fatal error message to the output stream.
      internal static void ReportFatalError(int lineNumber, Error error, 
         object arg0, object arg1, object arg2)
      {
         string message = string.Format(errorMap[error], arg0, arg1, arg2);
         InternalReportMessage(Output.error, lineNumber, message, (int)error);
         ErrorCount++;

         InternalReportMessage(Output.fatalError, lineNumber, 
            "unable to recover from previous error(s); stopping compilation.",
            (int)Error.FatalError);

         // Throw fatal error message.
         throw new FatalErrorException(lineNumber, message);
      }

      // Writer object for messages.      
      public static TextWriter MessageWriter;

      // The name of the file that is currently being compiled.
      public static string SourceFileName;

      // Retrieves the default system message writer.
      public static TextWriter DefaultWriter
      {
         get { return System.Console.Out; }
      }

      // The current error count.
      public static int ErrorCount;

      // The current warning count.
      static int WarningCount;

      static Output()
      {
         MessageWriter = null;
         
         ErrorCount = 0;
         WarningCount = 0;

         InitializeErrorMap();
      }

      // Internal method for writing strings to the output stream.
      private static void InternalReportMessage(
         string messageType, int lineNumber, string message, int messageNumber)
      {
         string output;

         // Write source file / line information if a valid line number
         // was provided.

         if (lineNumber >= 0)
         {
            output = String.Format("{0}({1}) ", SourceFileName, lineNumber);

            if (MessageWriter != null)
            {
               MessageWriter.Write(output);
            }
            else
            {
               Console.Write(output);
            }
         }

         // Add in the message number if one was provided.

         if (messageNumber >= 0)
            output = String.Format("{0} P{1}", messageType, messageNumber);
         else
            output = messageType;

         // Write the output.

         if (MessageWriter != null)
         {
            MessageWriter.Write(output);
         }
         else
         {
            if (messageType.Equals(error) || messageType.Equals(fatalError))
            {
               Console.ForegroundColor = System.ConsoleColor.Red;
            }
            Console.Write(output);
            Console.ResetColor();
         }

         output = String.Format(": {0}", message);

         if (MessageWriter != null)
         {
            MessageWriter.WriteLine(output);
         }
         else
         {
            Console.WriteLine(output);
         }
      }

      // Initializes the error map.
      private static void InitializeErrorMap()
      {
         errorMap = new Dictionary<Error, string>();

         errorMap[Error.UndeclaredIdentifier] =
            "'{0}': undeclared identifier";

         errorMap[Error.TooManyArguments] =
            "higher-order functions can have no more than {0} arguments.";

         errorMap[Error.FileNotFound] = 
            "'{0}': file not found.";
            
         errorMap[Error.InternalError] =
            "internal error: '{0}'.";
      }

      // Prefix strings for warnings and errors.

      private const string trace = "trace";
      private const string warning = "warning";
      private const string error = "error";
      private const string fatalError = "fatal error";

      // Maps Error values to their string representations.
      private static Dictionary<Error, string> errorMap;

      // Prepares the console for a block of output messages.
      internal static void BeginVerboseOutput()
      {
         Console.ForegroundColor = System.ConsoleColor.Yellow;         
      }

      // Resets console after a block of output messages.
      internal static void EndVerboseOutput()
      {
         Console.ResetColor();
      }
   }
}